### ISSUE TITLE
Command line building

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#508]
@mansellan suggested that tB could potentially used inside the OSS Rubberduck project (https://github.com/WaynePhillipsEA/twinbasic/discussions/383#discussioncomment-1707853).  Though CI would require the twinBASIC compiler to be able to build executables from the command line, which is not currently possible.

The syntax might be as simple as shown below;
`twinBASIC_win32.exe --build c:\MyProject.twinproj --out c:\out\MyProject.exe`
or 
`twinBASIC_win64.exe --build c:\MyProject.twinproj --out c:\out\MyProject.exe`

Note that error reporting also needs to be routed to the command line.

**Comment by mansellan** [comment]
Just a thought - you should probably allow in the license for license holders to install a single additional copy for unattended CI, otherwise it could cause problems for commercial use.

Not a lawyer, so I'll stop short of suggesting wording!

**Comment by mansellan** [comment]
Would also be nice to be able to specify conditional compilation arguments.

**Comment by EduardoVB** [comment]
And doing to back-compatible with VB6 command line options?

**Comment by WaynePhillipsEA** [comment]
Yes, I think that would be wise in the long term

**Comment by EduardoVB** [comment]
VB6.exe /? shows:

![imagen](https://user-images.githubusercontent.com/42319299/180053980-20be0467-117e-45be-8e80-7b656c789d96.png)

**Comment by rexxitall** [comment]
The fun part of VB6 is that you could not create  (non gui) executables. Even the compiler could do that. Ive long time used a hack to be able to do that sort of programs :) By all love to be as original as possible i suggest to have also a switch then to just compile a non gui programm without hacking twinbasic ;) In fact if i try to teach someone basic i make a wide circle around gui things first. For vary good reasons. ;) Usually if they are happy to master a for next they want to rule the universe with the next gui app noone needs. And they missed all the nice things what you can do withozt a gui. So in sound words nearly everything :)

**Comment by bclothier** [comment]
FYI - tB already support creating a console program -- it's a setting via the `Settings`. However, it's missing an equivalent `Console` object, so you need to use API for it. See #47 .

**Comment by rexxitall** [comment]
I know but you can not see its output in the IDE ;) Especially not my beautiful game *grin* 
![image](https://user-images.githubusercontent.com/22050310/180340291-375c1a05-6281-4e61-9ee9-b98eb03e8e1e.png)

**Comment by rexxitall** [comment]
Has i bit the charme of a c64 or c128. Til today i missed the feature just with the textmode chars to draw easy beautiful interfaces. The old comodore font was genius. You can had even colorful round edges boxes in mid of the 80ties. :) Til today not really be possible in text mode.[was maybe too easy to implemet] :D

**Comment by bclothier** [comment]
Just to note as I don't see it noted in the help dialog screenshot, we should also make sure to replicate the VB6's "command line" behavior at least for add-ins that may be used by the executable via command line. Not sure if it has applicability outside the context of an add-in. 
![image](https://user-images.githubusercontent.com/2367644/180365555-d9b6e8f9-1786-4021-a766-fc1800878b1e.png)

Also, I noticed that if the `/out` is not specified errors will be shown in a messagebox. Can we avoid this? (maybe that's what Wayne was alluding to in the last sentence in the OP) I always thought it was very weird that a messagebox would be used in response to a command. It's one thing to show messagebox when the user has opened it by double-clicking on it but from command line, it just feels wrong and would be a A Very Bad Thing™ in scenarios like CI.

**Comment by mansellan** [comment]
> Also, I noticed that if the /out is not specified errors will be shown in a messagebox. Can we avoid this? (maybe that's what Wayne was alluding to in the last sentence in the OP) I always thought it was very weird that a messagebox would be used in response to a command. It's one thing to show messagebox when the user has opened it by double-clicking on it but from command line, it just feels wrong and would be a A Very Bad Thing™ in scenarios like CI.

Definitely agree - this was a _very_ weird choice by MS. If you're building from the command line, build output should go to, uh, the command line...

**Comment by wqweto** [comment]
Unfortunately an EXE in Windows cannot be marked GUI subsystem and then handle console properly/seamlessly.

Definitely we need separate console application outside TB IDE so that it can be used in batch compilation scenarios only.

VB6 IDE parameters and support for batch compilation is a kludge IMO, a clear afterthought.

**Comment by WaynePhillipsEA** [comment]
That's very true @wqweto.  I've been bitten by that before. Luckily tB is already split into DLLs, so creating a separate console app for this would be quite straightforward.

**Comment by rexxitall** [comment]
Quite good with the dlls. I suggest to go a step further and let the engine run as a service / server. So it can live in memory active and clients may connect via tcpip or websocks or whatever. Websocks are nice. Ive written long ago a program for met@box whee we fetch stock market datas on that way in real time. Quite fast in OS/2 on a 200 mhz computer 1998. I never forget about that design and i use it for my Autocad thing. It does not matters in this case which OS the client runs or if he is a GUI guy or a command line window. Implementation is also rather easy and you keep the engine apart from the design stuff. If you provides a sound documentation of the interface you will be surprised how the peoples will use it :). here a picture from what i have so far.  Same engine different clients. 
![image](https://user-images.githubusercontent.com/22050310/180499900-dba79b6b-ecce-459d-92a4-59d2f1fddcf9.png)

**Comment by rexxitall** [comment]
And before you ask. Yes its some kind of LOGO Interpreter there. Makes much more fun if you has to deal with automatic polygon lines :). And you may me hit also a bit. Its written in VBS and can act as a CGI script inside Apache or inside Autocad native. bith has same code base.

**Comment by mansellan** [comment]
IIUC, running as a service would require an installer though, and it's really quite nice that twinBASIC doesn't need one atm.

Maybe it could be an advanced install option for those that need it?

**Comment by rexxitall** [comment]
All it needs is a VBS script if someones wants to have it.

**Comment by WaynePhillipsEA** [comment]
My preference would be to only allow command line building in the paid tiers of twinBASIC.  My reasoning for this is that if you require that level of integration then you are more than likely creating something commercial.  Any thoughts or objections?

**Comment by wqweto** [comment]
I've never heard of a compiler which cannot build from command line (for CI) in its community edition but of course you can do whatever keeps the project afloat.

Wondering if the free version of VB5 (Control Creation Edition or similar) can build OCXes from command line?

**Comment by rexxitall** [comment]
I do not think that this is a good idea. A lot of free software and eco systems live that someone can make automatic builds daily. So this will make it impossible. No one needs to write a damn driver :P on such things you can think of. I fully understand the Lambo / Porsche thing but how we say in Germany. Keep the church in the village. Best way to get paid customers is a great product, great service and a reasonable price. As far as i know like in the good old days you will just generate software piracy. Sooner or later whatever get hacked. Its not the question if - just when :) If i have to buy software in private then there are two criteria not to buy anything. Price over 200$ and price by annual fee (my limits). But for sure that is your point of decision. Pascal is dead by greed. No one use it anymore. On the other hand M$ is still in biz even the community edition of VS is for free. :)

**Comment by EduardoVB** [comment]
Usually the money comes from companies. Try yo get everybody onboard and when they are employees they say "we need to buy this".
Most companies do not even think on using illegal software (with the exception of some countries).
MS let pirate their products most probably willingly. But that's not good anyway, it is best to make an schema that lets people be in the right side of the law.
It is a difficult subject, I mean to come up with the best decisions.

**Comment by EduardoVB** [comment]
And about the subscription model, I as an user do not like it too much, but for companies it is the way to go. Otherwise you sell a license once and that's it (that's a problem that I'm having BTW).

**Comment by jpbro** [comment]
From a business perspective, I think a monthly subscription model for tB is required. As a programmer I'm happy with that provided I can compile *forever* with my last paid version (no warranty of course) without paying an extra cent.

**Comment by EduardoVB** [comment]
I thought that when a license was not renewed you couldn't compile anymore.

**Comment by jpbro** [comment]
> I thought that when a license was not renewed you couldn't compile anymore.
Yes, AFAIK that is the current license, though I'll admit that I only glanced it.  Once I tried compiling some of my code with the early and free tB releases (and it mostly worked!!), I knew that @WaynePhillipsEA was the real deal. I immediately said "Take my wallet!". Unfortunately for Wayne, my wallet isn't very full ;)

> If you can still compile when it expires then buying a year subscription is like buying a license for life. And what happens if the person cancels the subscription after a month?

I could only compile against the last version I subscribed to. I've paid for licenses for a slew of ActiveX libraries/controls, and almost all of them had this model:

1) Royalty Free
2) Free Updates during the subscription period, after that no updates (but you can still compile against the most recent version you subscribed to).

IMO, this is the most fair model. As a "consumer" of Wayne's work, I'll pay him in perpetuity provided I am seeing value returned. If I no longer see value returned, I'll stop paying - but I won't lose the ability to do what I was able to do yesterday.

**Comment by EduardoVB** [comment]
I don't think that's fair for the developer.
You pay one month and then have the ability to use the product forever.
For the same thing MS for VB6 asked for a full license. They didn't issue updates to you either (well, some SP that were bug fixes).

PS: you posted your first answer above as you were quoting me.

**Comment by jpbro** [comment]
Sorry about the answer/quote thing- I didn't press Enter twice, so GitHub assumed it was all one block.

Regarding fairness, let's look at the other side. The developer releases something awesome and charges $X/month. I buy in and build my business around it, but then they never update it again. Eventually there are bugs/issues/bit-rot, but the developer no longer cares to update the software. Should I keep paying to be able to compile?

Re: MS - I paid ~$300 back in 1998 (~$12/*year*! if we look at it at as a "subscription"). I can still compile VB6 apps to this day. Even though there have only been security updates in the interim, I wouldn't be surprised if MS spends more than $200 a day to make sure that VB6 compiled apps work on all OS releases since then.

Another example - I paid ~ $400-600/year to Codejock for a few releasesof their ActiveX suite. I did this from version 11.x-16.x if I recall correctly. I stopped subscribing after 16.x because I didn't see any new value in the subscription, but I can still compile against the last subscribed version to this day. They released 20.x with improved HighDPI support, and I resubscribed for a year. Now I can compile against 20.x. If future versions don't add value, then I won't renew my subscription. If they do, then I will. This seems fair - it keeps the developer working instead of rent-seeking.

Another example is ComponentOne - I purchased as yearly subscription to their full 2004 suite (I was exploring .Net at the time) for around ~1000 if I recall correctly. They completely abandoned ActiveX control development a couple of years later, but I can still compile against the 2004 version. If I install the 2005 or 2006 version though, I can't compile against that - which is totally fair, because I didn't pay for that.

**Comment by xiaoyaocode163** [comment]
can twinbasic make exe by command line ,only use vb6 project ?

**Comment by fafalone** [comment]
Not currently but that would be an interesting feature... once command line compilation for .twinproj is implemented, it would make sense to offer an option to automatically import and compile. Though I'm not sure how common the use case is... if you're not upgrading the project to tB, why do you want to compile with it?

But I'd support adding such an option as part of having a robust command line like other compilers.

**Comment by GaborNyul** [comment]
Hi All, in the past years have you managed to progress with the command line building of existing twinbasic projects? Purpose would be to make twinbasic projects part of a CI/CD pipeline.

**Comment by wqweto** [comment]
How do you plan to automate CI/CD pipelines when .twinproj files are opaque binaries? How do you plan to spin off CI on source file commits when you currently cannot stuff these into .twinproj files?

This issue should probably get flagged to depend on the already existing `Support external source files in project files` issue in first place.

**Comment by GaborNyul** [comment]
> How do you plan to automate CI/CD pipelines when .twinproj files are opaque binaries? How do you plan to spin off CI on source file commits when you currently cannot stuff these into .twinproj files?

Put everything including .twinproj files into git. CI/CD pulls repo in case there is an update. Calls twinbasic (local onprem, vm or docker) and uses the produced exe file to run tests.

**Comment by wqweto** [comment]
Possible though sounds awkward. (Why use version control on blobs when you cannot diff anything.)

Hope we get linked source files before command-line compilation although both seem like low-hanging fruits. The letter will most probably need a separate CLI tool/driver to run the actual compiler.

**Comment by GaborNyul** [comment]
> Possible though sounds awkward. (Why use version control on blobs when you cannot diff anything.)

For the very same reason why you are not introducing Kubernetes for an app that can be first vertically scaled. Small scale with small files like a twinbasic project file the simplest choice to store in git with the rest of your code. That way your CI/CD will remain simple solution. On a large scale, you can use, e.g., a binary repository. KISS principle in a way.
I am happy to discuss the pros and cons of it privately, so we can keep this topic clean.

**Comment by fafalone** [comment]
In any case, there's a preliminary version of command line compiling now. 

`twinbasic.exe PathToProject.twinproj --buildAndExit32` 

or ..64 for 64bit. No error handling etc, just experimental.

**Comment by KubaO** [comment]
> How do you plan to automate CI/CD pipelines when .twinproj files are opaque binaries? How do you plan to spin off CI on source file commits when you currently cannot stuff these into .twinproj files?

.twinprojs are hardly opaque. Here's a [010 Editor][1] template that decodes them:

```C
typedef struct {
    uint length <bgcolor=0x008000, hidden=true>;
    if (length > 0)
        uchar value[length] <bgcolor=0x800080>;
} String <read = Str("0x%04x \"%s\"", this.length, this.value)>;

local int entryCount = 0;

struct AnEntry_;
typedef struct AnEntry_ {
    short kind <hidden=true>;
    String name;
    ushort mark1 <hidden=true>;
    uchar pad[10] <hidden=true>;
    uchar mark2 <hidden=true>;
    entryCount += 1;
    if (kind == 1 && entryCount > 1) {
        String contents;
        uint trailer <hidden = true>;
    } else {
        uint count <hidden=true>;
        if (count > 0)
            AnEntry_ entries[count] <read = readEntry_, optimize=false>; 
    }
} AnEntry <read = readEntry>;

string readEntry_(struct AnEntry_& entry) {
    return Str("%d %04x %02x %s", entry.kind, entry.mark1, entry.mark2, entry.name.value);
}

string readEntry(AnEntry& entry) {
    return Str("%d %04x %02x %s", entry.kind, entry.mark1, entry.mark2, entry.name.value);
}

int magic <bgcolor=0x808000, format=hex>;

AnEntry top <optimize=false>;
```

`mark1` is the file modification counter. Each time a new version of a file is saved, it increments by 2. It starts at 2. It is zero for folders. `mark2` indicates a folder type or 0 for files.

There is enough known about this binary file to round-trip it via a series of folders and text files, i.e. an export and import, using a small script, so git will be quite happy with it.

For inspiration see e.g. https://bitbucket.org/sippey/zippey/src/master/zippey.py

I'll probably write a git filter to deal with .twinprojs soon.

[1]: https://www.sweetscape.com/010editor/

**Comment by listhexagon** [comment]
@KubaO a Git filter to round-trip the .twinproj files sounds like a really good idea.
Is there any progress on creating such a filter?

**Comment by KubaO** [comment]
@listhexagon It should be coming within 4 weeks. As a stopgap, I have revamped the impexp tool to support tB's own import/export command line syntax. I have also added a [git-specific section to the package import/export page][1].

[1]: https://docs.twinbasic.com/Features/Packages/Import-Export-Tool#keeping-a-project-in-git

**Comment by KubaO** [comment]
In the documentation build system, I had to implement a way to build and run tB projects. The tools are:

- [`tbbuild`][1]

   The build tool currently does not support writing an EXE. It will be an option I'll add. I currently use the tool to check if a given project builds cleanly, while not needing the `.exe` result for anything.

- [`tbrun`][2] is *not useful* for running projects in general yet; it has a diagnostic purpose when comparing documented behavior with observed behavior.

These tools are under the MIT license that all of the documentation tooling is under, and full JS source code is included. To use the tools, clone https://github.com/twinbasic/documentation.git and then run `npm install` in the root of the cloned repository.

[1]: https://docs.twinbasic.com/Documentation/Development/Tools#tbbuild
[2]: https://docs.twinbasic.com/Documentation/Development/Tools#tbrun
